src/zitadel.rs

```diff
@@ -47,16 +47,46 @@ impl Zitadel {
 			zitadel_client,
 		})
 	}
+
 	/// Import a list of new users into Zitadel
 	pub(crate) async fn import_new_users(&self, users: Vec<User>) -> Result<()> {
 		for user in users {
 			let zitadel_user =
 				user.to_zitadel_user(&self.feature_flags, &self.zitadel_config.idp_id);
-			let sync_status = self.import_user(&zitadel_user).await;
+			let status = self.import_user(&zitadel_user).await;
 
-			if let Err(error) = sync_status {
-				tracing::error!("Failed to sync user `{}`: {:?}", zitadel_user.log_name(), error);
-			};
+			if let Err(error) = status {
+				tracing::error!(
+					"Failed to sync-import user `{}`: {:?}",
+					zitadel_user.log_name(),
+					error
+				);
+
+				if Self::is_invalid_phone_error(error) {
```

Comment:
The code is still incredibly similar between the two functions, I think there's a single word difference between the two implementations. Can't this logic be shared too?

Personally I wonder if this logic should be lifted up one abstraction layer, returning an `Error` with an enum value we can process in the parent function call, which then rearranges another set of imports with adjusted phone numbers. Not that I've fully thought through a good architecture for this, but needing this many extremely specific utility functions feels like a code smell.

I suppose it's fine to hold off on thinking through things too much for now since we do need this feature, but we should redesign this in the future.

Reply:
From what I understand, you are trying not go against DRY.

Let's think about it in terms of pseudocode. Currently:

1. ask zitadel-rust-client to **import** a user
2. call the local function to check if returned error is about phone numbers
3. if so, try to **import** the user without phone number

Then the other place we go over this procedure, instead of **import** it is **update**. So the only step that repeats is stripping the user of the phone number.

Looks clean enough to me - at least at this point. If we do decide to redesign this in the future, I will be more than happy to assist if I can.